cargo-miri/bin.rs

```diff
@@ -251,26 +251,72 @@ fn xargo_check() -> Command {
 
 /// Execute the command. If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
-fn exec(mut cmd: Command) {
-    let exit_status = cmd.status().expect("failed to run command");
-    if exit_status.success().not() {
-        std::process::exit(exit_status.code().unwrap_or(-1))
+fn exec(mut cmd: Command) -> ! {
+    // On non-Unix imitate POSIX exec as closely as we can
+    #[cfg(not(unix))]
+    {
+        let exit_status = cmd.status().expect("failed to run command");
+        std::process::exit(exit_status.code().unwrap_or(-1));
+    }
+    // On Unix targets, actually exec
+    // If exec returns, process setup has failed. This is the same error condition as the expect in
+    // the non-Unix case.
+    #[cfg(unix)]
+    {
+        use std::os::unix::process::CommandExt;
+        let error = cmd.exec();
+        Err(error).expect("failed to run command")
     }
 }
 
 /// Execute the command and pipe `input` into its stdin.
 /// If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
 fn exec_with_pipe(mut cmd: Command, input: &[u8]) {
-    cmd.stdin(process::Stdio::piped());
-    let mut child = cmd.spawn().expect("failed to spawn process");
+    #[cfg(unix)]
     {
-        let stdin = child.stdin.as_mut().expect("failed to open stdin");
-        stdin.write_all(input).expect("failed to write out test source");
+        use std::os::unix::io::FromRawFd;
+        use std::process::Stdio;
+
+        let mut fds: [libc::c_int; 2] = [0, 0];
+        // SAFETY: fds is an array of 2 libc::c_int
+        let res = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        assert_eq!(res, 0, "failed to create pipe");
+
+        // We need to set close-on-exec, otherwise our pipe isn't readable after exec
+        // SAFETY: fcntl has no preconditions
+        unsafe {
+            for fd in &fds {
+                let res = libc::fcntl(
+                    *fd,
+                    libc::F_SETFD,
+                    libc::fcntl(*fd, libc::F_GETFD) | libc::FD_CLOEXEC,
+                );
+                assert_eq!(res, 0, "failed to set close-on-exec for pipe");
+            }
+        }
+
+        // SAFETY: Both elements of fds are open file descriptors, because pipe2 returned 0
+        let dst = unsafe { Stdio::from_raw_fd(fds[0]) };
+        let mut src = unsafe { File::from_raw_fd(fds[1]) };
+
+        src.write_all(input).expect("failed to write out test source");
+
+        cmd.stdin(dst);
+        exec(cmd)
```

Comment:
So on Unix this never returns, but on Windows it does? That doesn't seem good.

Reply:
Windows diverges too, I just didn't make that clear before. Now I think it is.